bottlerocket/agents/src/bin/ecs-workload-agent/main.rs

```diff
@@ -0,0 +1,563 @@
+            ),
+        )
+        .await
+        {
+            Ok(results) => results,
+            Err(_) => {
+                let mut res =
+                    test_results(&ecs_client, &self.config.cluster_name, &task_arns).await;
+
+                // We've timed out, but there is a window where our final check could actually report completion. To
+                // guard against this, check if still in progress and marked failed only if so.
+                if let Ok(mut result) = res {
+                    if result.outcome == Outcome::InProgress {
+                        result.outcome = Outcome::Fail;
+                    }
+                    res = Ok(result);
+                }
+
+                res
+            }
+        }
+    }
+
+    /// Terminate the test runs. This is a noop for this agent.
+    async fn terminate(&mut self) -> Result<(), Self::E> {
+        Ok(())
+    }
+}
+
+/// Loops waiting for the workload test tasks to complete.
+async fn wait_for_test_completion<I>(
+    ecs_client: &aws_sdk_ecs::Client,
+    cluster_name: &str,
+    task_arns: &[String],
+    info_client: &I,
+) -> Result<TestResults, Error>
+where
+    I: InfoClient,
+{
+    let mut retries: u64 = 0;
+    loop {
+        let mut results = test_results(ecs_client, cluster_name, task_arns).await?;
+        if results.outcome == Outcome::Pass {
+            info_client
+                .send_test_update(results.clone())
+                .await
+                .err()
+                .iter()
+                .for_each(|e| log::error!("Unable to send test update: {}", e));
+            return Ok(results);
+        }
+
+        retries += 1;
+        results.other_info = Some(format!("Waiting for test completion: attempt {}", retries));
+        info_client
+            .send_test_update(results.clone())
+            .await
+            .err()
+            .iter()
+            .for_each(|e| log::error!("Unable to send test update: {}", e));
+        tokio::time::sleep(Duration::from_secs(2)).await;
+    }
+}
+
+/// Gets the status from our test tasks and populates a `TestResults` struct with the outcome.
+async fn test_results(
+    ecs_client: &aws_sdk_ecs::Client,
+    cluster_name: &str,
+    task_arns: &[String],
+) -> Result<TestResults, Error> {
+    let tasks = ecs_client
+        .describe_tasks()
+        .cluster(cluster_name)
+        .set_tasks(Some(task_arns.to_vec()))
+        .send()
+        .await
+        .context(error::TaskDescribeSnafu)?
+        .tasks()
+        .map(|tasks| tasks.to_owned())
+        .context(error::NoTaskSnafu)?;
+    let running_count = tasks
+        .iter()
+        .filter(|task| task.last_status() == Some("STOPPED"))
+        .filter(|task| task.stop_code() == Some(&TaskStopCode::EssentialContainerExited))
+        .filter(|task| {
+            task.containers()
+                .unwrap_or_default()
+                .iter()
+                .filter(|container| container.exit_code() != Some(0))
```

Comment:
Should we count on failed tasks so we can accurately report them in test results? I don't think we should mark a test as a failure if it hasn't finished executing.

Reply:
In this case, `num_passed + num_failed < task_count` unless all tests completed.

Reply:
We should report failed tasks in the status updates, but wait until all tasks finish before reporting results.

Reply:
I like that better. 👍